### PR TITLE
Implement MySQL time service

### LINKED ISSUE #3326: Implement TimeService for MySQL
## Feature Request

**For English only**, other languages will not accept.

Please pay attention on issues you submitted, because we maybe need more details. 
If no response **more than 7 days** and we cannot make decision by current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Is your feature request related to a problem?
Yes
### Describe the feature you would like.
Using class Implementing TimeService to execute `select now()`  for MySQL, determine time service DataSource through configuration
but I have two problems
1. Which module should this class be placed on?
2. How to give this class the ability to execute

### PR DESCRIPTION
**Pull request by Pramy**
Change-Id: I791c1d3ff4c4cbf0f90464c7bb8bb9490dca4019

Fixes #3326 .

Changes proposed in this pull request:
-
-
-

### CODE REVIEW

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    static {
+        init();
+    }
+
+    @Override
+    public Date getTime() {
+        if (null != dataSource) {
+            try (Connection connection = dataSource.getConnection()) {
+                Statement statement = connection.createStatement();
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java]: Please use PreparedStatement to improvement performance

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                resultSet.next();
+                return (Date) resultSet.getObject(1);
+            } catch (SQLException ignore) {
+
+            }
+        }
+        return new Date();
+    }
+
+    private static void init() {
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java]: Please move this method in front of `getTime`, we should keep same method's order with the call's order

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import java.lang.reflect.Method;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Date;
+import java.util.Properties;
+
+/**
+ * Time service for SPI.
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java]: Please fix javadoc

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        }
+        return new Date();
+    }
+
+    private static void init() {
+        try {
+            Properties properties = new Properties();
+            properties.load(MySQLTimeService.class.getResourceAsStream("/mysql-time-service.properties"));
+            if (properties.isEmpty()) {
+                return;
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java]: Please throw exception if necessary missing information

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    @Override
+    public Date getTime() {
+        if (null != dataSource) {
+            try (Connection connection = dataSource.getConnection()) {
+                Statement statement = connection.createStatement();
+                ResultSet resultSet = statement.executeQuery("select now()");
+                resultSet.next();
+                return (Date) resultSet.getObject(1);
+            } catch (SQLException ignore) {
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java]: Please add final for exception declare

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            String dataSourceType = (String) properties.remove("dataSourceType");
+            Class dataSourceClass = Class.forName(dataSourceType);
+            DataSource dataSource = (DataSource) dataSourceClass.newInstance();
+            for (String key : properties.stringPropertyNames()) {
+                PropertyDescriptor propDesc = new PropertyDescriptor(key, dataSourceClass);
+                Method writeMethod = propDesc.getWriteMethod();
+                writeMethod.invoke(dataSource, properties.getProperty(key));
+            }
+            MySQLTimeService.dataSource = dataSource;
+        } catch (IOException | ClassNotFoundException | InstantiationException | IllegalAccessException | InvocationTargetException | IntrospectionException ignore) {
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java]: Please add final for exception declare

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        try {
+            Properties properties = new Properties();
+            properties.load(MySQLTimeService.class.getResourceAsStream("/mysql-time-service.properties"));
+            if (properties.isEmpty()) {
+                return;
+            }
+            String dataSourceType = (String) properties.remove("dataSourceType");
+            Class dataSourceClass = Class.forName(dataSourceType);
+            DataSource dataSource = (DataSource) dataSourceClass.newInstance();
+            for (String key : properties.stringPropertyNames()) {
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java]: Please use `each` as variable name in a loop

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+                PropertyDescriptor propDesc = new PropertyDescriptor(key, dataSourceClass);
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java]: Please do not use abbreviation for `propDesc`, `propertyDescriptor` is better

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java (hunk 6 above)
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/MySQLTimeService.java]: Need to throw exception if error occur

File: sharding-core/sharding-core-route/src/test/java/org/apache/shardingsphere/core/route/spi/MySQLTimeServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+package org.apache.shardingsphere.core.route.spi;
+
+import lombok.SneakyThrows;
+import org.junit.Assert;
+import org.junit.Test;
+
+import javax.sql.DataSource;
+import java.lang.reflect.Field;
+
+public class MySQLTimeServiceTest {
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/test/java/org/apache/shardingsphere/core/route/spi/MySQLTimeServiceTest.java]: Should add final if class do not for extends

**Comment by Pramy** [comment]
> Please fix all code review suggestions.
> 
> And more more question for design, how to enable this SPI? Does need user add to SPI file by themselves?
> Maybe we should consider about spilt this code to a new module and setup SPI fine in the new module.
> If user import this module, they can load from SPI automatically.
> 
> For the default time service , maybe we should consider about do not use SPI to load it. The logic maybe better `if no other SPI, just use default one`.
> How do you think?

that sound great. user can controll whether they need to use `MySQLTimeService`. if do that. what is the module name？ and where should the module place on？

**Comment by terrymanu** [comment]
> > Please fix all code review suggestions.
> > And more more question for design, how to enable this SPI? Does need user add to SPI file by themselves?
> > Maybe we should consider about spilt this code to a new module and setup SPI fine in the new module.
> > If user import this module, they can load from SPI automatically.
> > For the default time service , maybe we should consider about do not use SPI to load it. The logic maybe better `if no other SPI, just use default one`.
> > How do you think?
> 
> that sound great. user can controll whether they need to use `MySQLTimeService`. if do that. what is the module name？ and where should the module place on？

The name maybe as `database-time-service`, and we can put it below route module?

**Comment by Pramy** [comment]
> > > Please fix all code review suggestions.
> > > And more more question for design, how to enable this SPI? Does need user add to SPI file by themselves?
> > > Maybe we should consider about spilt this code to a new module and setup SPI fine in the new module.
> > > If user import this module, they can load from SPI automatically.
> > > For the default time service , maybe we should consider about do not use SPI to load it. The logic maybe better `if no other SPI, just use default one`.
> > > How do you think?
> > 
> > 
> > that sound great. user can controll whether they need to use `MySQLTimeService`. if do that. what is the module name？ and where should the module place on？
> 
> The name maybe as `database-time-service`, and we can put it below route module?

Ths. I finished

File: sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/exception/TimeServiceInitException.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+package org.apache.shardingsphere.route.time.exception;
+
+/**
+ * TimeService init exception.
+ *
+ * @author chenchuangliu
+ */
+public class TimeServiceInitException extends RuntimeException {
```
Comment (terrymanu) [inline sharding-core/database-time-service/src/main/java/org/apache/shardingsphere/route/time/exception/TimeServiceInitException.java]: please add final if class is not for extend

**Pull request merged**